src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -693,46 +700,130 @@ private unsafe int FindFirstCharacterToEncode(ReadOnlySpan<char> text)
         /// current encoder instance, or -1 if no data in <paramref name="utf8Text"/> requires escaping.
         /// </returns>
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public virtual int FindFirstCharacterToEncodeUtf8(ReadOnlySpan<byte> utf8Text)
+        public virtual unsafe int FindFirstCharacterToEncodeUtf8(ReadOnlySpan<byte> utf8Text)
         {
-            int originalUtf8TextLength = utf8Text.Length;
-
             // Loop through the input text, terminating when we see ill-formed UTF-8 or when we decode a scalar value
             // that must be encoded. If we see either of these things then we'll return its index in the original
             // input sequence. If we consume the entire text without seeing either of these, return -1 to indicate
             // that the text can be copied as-is without escaping.
 
-            int i = 0;
-            while (i < utf8Text.Length)
+            fixed (byte* ptr = utf8Text)
             {
-                byte value = utf8Text[i];
-                if (UnicodeUtility.IsAsciiCodePoint(value))
+                int idx = 0;
+
+#if NETCOREAPP
+                if (Sse2.IsSupported)
                 {
-                    if (!ReferenceEquals(GetAsciiEncoding(value), s_noEscape))
+                    sbyte* startingAddress = (sbyte*)ptr;
+                    while (utf8Text.Length - 16 >= idx)
                     {
-                        return originalUtf8TextLength - utf8Text.Length + i;
+                        Debug.Assert(startingAddress >= ptr && startingAddress <= (ptr + utf8Text.Length - 16));
+
+                        // Load the next 16 bytes.
+                        Vector128<sbyte> sourceValue = Sse2.LoadVector128(startingAddress);
+
+                        Vector128<sbyte> mask = Sse2Helper.CreateAsciiMask(sourceValue);
+                        int index = Sse2.MoveMask(mask.AsByte());
+
+                        int processNextSixteen = idx + 16;
+                        Debug.Assert(processNextSixteen <= utf8Text.Length);
+                        if (index != 0)
+                        {
```

Comment:
Move this block into a separate method, to keep the scan-loop tight?

Reply:
Moving this out to another method slows things down quite a bit.

Reply:
The culprit here was passing idx by ref.